main_IK.py

```diff
@@ -0,0 +1,160 @@
+amp_z = 0.1 
+freq_x = 0.1
+freq_y = 0.2 
+freq_z = 0.1
+
+body_name = "wrist_3_link"
+body_id = model.body(body_name).id
+jacp = np.zeros((3, model.nv))
+jacr = np.zeros((3, model.nv))
+
+with mujoco.viewer.launch_passive(model, data) as viewer:
+    for _ in range(int(duration / model.opt.timestep)):
+        t = data.time - sim_start
+        
+        J_pos = np.zeros(3 * model.nv)
+        J_rot = np.zeros(3 * model.nv)
+
+        target_pos = base_target + np.array([amp_x * np.sin(2 * np.pi * freq_x * t),amp_y * np.sin(2 * np.pi * freq_y * t),amp_z * np.sin(2 * np.pi * freq_z * t)])
+        
+        q_des = ik_solver.solve(target_pos, body_name="wrist_3_link", q_init=data.qpos[:model.nq])
+        
+        x_vel = np.array([amp_x*(2*np.pi*freq_x)*np.cos(2 * np.pi * freq_x * t), amp_y*(2*np.pi*freq_y)*np.cos(2 * np.pi * freq_y * t), amp_z*(2*np.pi*freq_z)*np.cos(2 * np.pi * freq_z * t)])
+        x_acc = np.array([-amp_x *(2*np.pi*freq_x)**2*np.sin(2 * np.pi * freq_x * t),-amp_y*(2*np.pi*freq_y)**2 * np.sin(2 * np.pi * freq_y * t), -amp_z*(2*np.pi*freq_z)**2 * np.sin(2 * np.pi * freq_z * t)])
+        
+        x_current = data.body(body_id).xpos.copy()
+        
+        mujoco.mj_jac(model, data, jacp, jacr, x_current, body_id)  
+        J = jacp[:, :model.nv]
+        
+        qd_des = np.linalg.pinv(J) @ x_vel  
+        qdd_des = np.linalg.pinv(J) @ x_acc
+        
+        data.qpos[:num_joints] = q_des
+        data.qvel[:num_joints] = qd_des
+        data.qacc[:num_joints] = qdd_des  
+        
+        mujoco.mj_inverse(model, data)  
+        tau = data.qfrc_inverse[:num_joints].copy()    
+        
+        # noise terms; not using this rn (can change this to disturbance or parameter uncertainty to make feedback control more interesting)
+        # noise_q = np.random.normal(0, 0.005, size=num_joints)  
+        # noise_qd = np.random.normal(0, 0.01, size=num_joints) 
+
+        q_meas = data.qpos[:num_joints] 
+        qd_meas = data.qvel[:num_joints] 
+        q_err = q_des - q_meas
+        qd_err = qd_des - qd_meas
+        integral_error += q_err * model.opt.timestep
+
+        # Feedforward + PI controller
+        data.ctrl[:num_joints] = tau + Kp * q_err + Ki * integral_error
+
+        # Log
+        tau_history.append(tau.copy())
+        q_des_history.append(q_des.copy())
+        qd_des_history.append(qd_des.copy())
+        q_act_history.append(q_meas.copy())
+        qd_act_history.append(qd_meas.copy())
+        
+        time_history.append(t)
+
+        mujoco.mj_step(model, data)
+        viewer.sync()  
+
+# After simulation: convert lists to arrays
+tau_history = np.array(tau_history)
+rows, columns = tau_history.shape
+print(f"Matrix dimensions: {rows} rows, {columns} columns")
+q_des_history = np.array(q_des_history)
+qd_des_history = np.array(qd_des_history)
+q_act_history = np.array(q_act_history)
+qd_act_history = np.array(qd_act_history)
+time_history = np.array(time_history)
+
+# Plot results
```

Comment:
As we discussed, I would recommend adding forward kinematics to this example so you can see how well the robot tracks Cartesian position rather than just joints. In other words, this example tracks the joint positions well, but can you verify that this translates to tracking the original position profile well? 

But for educational purposes, I'm going to suggest you make a separate issue for that and we can address in a separate PR.